consumerOptions.go

```diff
@@ -21,51 +21,54 @@
 package kafkaclient
 
 import (
-	"github.com/Shopify/sarama"
-	"github.com/bsm/sarama-cluster"
 	"github.com/uber-go/kafka-client/internal/consumer"
-	"time"
 )
 
-// ConsumerOption is the type for optional arguments to the NewConsumer constructor.
 type (
+	// ConsumerOption is the type for optional arguments to the NewConsumer constructor.
 	ConsumerOption interface {
 		apply(*consumer.Options)
 	}
-	consumerLimitOption struct {
-		limits map[consumer.TopicPartition]int64
+
+	consumerBuildOptions []ConsumerOption
+
+	partialConstruction struct {
+		enabled bool
+		errs    *consumerErrorList
 	}
 )
 
-var defaultOptions = consumer.Options{
-	Concurrency:            1024,
-	RcvBufferSize:          2 * 1024, // twice the concurrency for compute/io overlap
-	PartitionRcvBufferSize: 32,
-	OffsetCommitInterval:   time.Second,
-	RebalanceDwellTime:     time.Second,
-	MaxProcessingTime:      250 * time.Millisecond,
-	OffsetPolicy:           sarama.OffsetOldest,
-	ConsumerMode:           cluster.ConsumerModePartitions,
-	Limits:                 nil,
+// PartialConstructionError returns a list of topics that could not be consumed as a list of ConsumerError.
+// PartialConstructionError should be called on the ConsumerOption returned by EnablePartitionConstruction.
+// This is useful if you chose to EnablePartialConstruction.
+func PartialConstructionError(option ConsumerOption) []ConsumerError {
+	pe, ok := option.(*partialConstruction)
+	if !ok {
+		return nil
+	}
+
+	return pe.errs.errs
 }
 
-// WithConsumerLimits sets consumer limits for a consumer.
-// If consumer limits are set, the consumer will only consume messages from the specified topic-partitions
-// up to the limits then shut itself down.
-// If limits are set and the consumer is assigned a topic-partition that is not in the limits map, no messages
-// will be received for that topic partition.
-func WithConsumerLimits(limits map[string]map[int32]int64) ConsumerOption {
-	topicPartitionLimits := make(map[consumer.TopicPartition]int64)
-	for topic, partitionMap := range limits {
-		for partition, offset := range partitionMap {
-			topicPartitionLimits[consumer.TopicPartition{Topic: topic, Partition: partition}] = offset
-		}
-	}
-	return &consumerLimitOption{
-		limits: topicPartitionLimits,
+// EnablePartialConstruction will set the client to return a partial consumer that
+// consumes from as many topics/clusters as it could and it may return an error that lists the
+// topics that failed to connect to their cluster.
+// You can use ConsumerBuildError(err error) to most information about the failed topics, if any.
+func EnablePartialConstruction() ConsumerOption {
```

Comment:
Consider renaming this to AllowPartialConsumption or EnablePartialConsumption. Also call out that this option only matters when consuming from topics spread across multiple clusters

Reply:
I went with EnablePartialConsumption.